server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorRuleRunner.java

```diff
@@ -77,13 +84,46 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
       return params;
     }
 
+    // find available segments which are not overshadowed by other segments in DB
+    // only those would need to be loaded/dropped
+    // anything overshadowed by served segments is dropped automatically by DruidCoordinatorCleanupOvershadowed
+    Map<String, VersionedIntervalTimeline<String, DataSegment>> timelines = new HashMap<>();
+    for (DataSegment segment : params.getAvailableSegments()) {
+      VersionedIntervalTimeline<String, DataSegment> timeline = timelines.get(segment.getDataSource());
+      if (timeline == null) {
+        timeline = new VersionedIntervalTimeline<>(Comparators.comparable());
+        timelines.put(segment.getDataSource(), timeline);
+      }
+
+      timeline.add(
+          segment.getInterval(), segment.getVersion(), segment.getShardSpec().createChunk(segment)
+      );
+    }
+
+    Set<DataSegment> overshadowed = new HashSet<>();
+    for (VersionedIntervalTimeline<String, DataSegment> timeline : timelines.values()) {
+      for (TimelineObjectHolder<String, DataSegment> holder : timeline.findOvershadowed()) {
+        for (DataSegment dataSegment : holder.getObject().payloads()) {
+          overshadowed.add(dataSegment);
+        }
+      }
+    }
```

Comment:
note that in DruidCoordinatorRuleRunner, we are considering a segment overshadowed if it is overshadowed by the segments available in DB.
and in DruidCoordinatorCleanupOvershadowed, we consider a segment overshadowed only if it is overshadowed by segments served from historicals (so that we don't prematurely mark something unused and drop it from historicals before the newer version is served).

the flow, you suggested, can be made to work but requires rethinking it all and I would be worried about introducing bugs. Given the sensitive nature of code here, I wanted to minimize the changes while addressing the problem. I think, current flow of things is ok and we can change it in a separate PR if/when needed.